src/platform/tizen/iotjs_tizen_service_app.c

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <dlog.h>
+#include <service_app_internal.h>
+#include <stdio.h>
+
+#include "iotjs_def.h"
+#include "iotjs.h"
+
+extern bool iotjs_jerry_init(iotjs_environment_t* env);
+extern int iotjs_start(iotjs_environment_t* env);
+extern void iotjs_uv_walk_to_close_callback(uv_handle_t* handle, void* arg);
+
+static char js_absolute_path[128];
+
+static int console_log(int level, const char* format, ...) {
+  va_list args;
+  va_start(args, format);
+  dlog_vprint(DLOG_INFO, "IOTJS", format, args);
+  va_end(args);
+  return 0;
+}
+
+static void loop_method_init_cb(int argc, char** argv, void* data) {
+  int iotjs_argc = 2;
+  char* iotjs_argv[2] = { "iotjs", js_absolute_path };
+
+  iotjs_debuglog_init();
+
+  iotjs_environment_t* env = iotjs_environment_get();
+  if (!iotjs_environment_parse_command_line_arguments(env, (uint32_t)iotjs_argc,
+                                                      iotjs_argv)) {
+    DLOG("iotjs_environment_parse_command_line_arguments failed");
+    service_app_exit();
+  }
+
+  if (!iotjs_jerry_init(env)) {
+    DLOG("iotjs_jerry_init failed");
+    service_app_exit();
+  }
+
+  iotjs_conf_console_out(console_log);
+}
+
+static void loop_method_run_cb(void* data) {
+  iotjs_environment_t* env = iotjs_environment_get();
+
+  // Set event loop.
+  iotjs_environment_set_loop(env, uv_default_loop());
+
+  // Start IoT.js.
+  iotjs_start(env);
+
+  service_app_exit();
+}
+
+static void loop_method_exit_cb(void* data) {
+}
+
+static void loop_method_fini_cb(void) {
+  iotjs_environment_t* env = iotjs_environment_get();
+
+  // Close uv loop.
+  uv_walk(iotjs_environment_loop(env), iotjs_uv_walk_to_close_callback, NULL);
+  uv_run(iotjs_environment_loop(env), UV_RUN_DEFAULT);
+
+  int res = uv_loop_close(iotjs_environment_loop(env));
+  IOTJS_ASSERT(res == 0);
+
+  // Release builtin modules.
+  iotjs_module_list_cleanup();
+
+  // Release JerryScript engine.
+  jerry_cleanup();
+
+  // Release environment.
+  iotjs_environment_release();
+
+  iotjs_debuglog_release();
+}
+
+int iotjs_service_app_start(int argc, char** argv, char* js_path,
+                            void* event_callbacks, void* user_data) {
+  char* app_res_path = app_get_resource_path();
+  if (!app_res_path) {
+    DLOG("app_res_path is NULL!");
+    return 1;
+  }
+
+  snprintf(js_absolute_path, sizeof(js_absolute_path), "%s%s", app_res_path,
+           js_path);
+  IOTJS_RELEASE(app_res_path);
+
+  service_app_loop_method_s loop_method = {.init = loop_method_init_cb,
+                                           .run = loop_method_run_cb,
+                                           .exit = loop_method_exit_cb,
+                                           .fini = loop_method_fini_cb };
+
+  return service_app_main_ext(argc, argv, (service_app_lifecycle_callback_s*)
```

Comment:
i) `service_app_main_ext` seems an undocumented api. If so, can you put some brief comments on this?
ii) Can you clarify when `loop_method_fini_cb` is called? Here `service_app_exit()` is used to exit the main loop. And, the reference says:

```
void service_app_exit  (void)

Exits the main loop of the application. 
`service_app_terminate_cb()` is invoked.
```
However we don't use `service_app_terminate_cb()` callback. 

https://developer.tizen.org/development/api-references/native-application?redirect=/dev-guide/2.3.1/org.tizen.native.mobile.apireference/group__CAPI__SERVICE__APP__MODULE.html#gae143970b24e8a00b519cafca85052079